src/core/common/const_cast.hpp

```diff
@@ -0,0 +1,103 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes helper methods to cast between const and non-const objects and/or pointers.
+ */
+
+#ifndef CONST_CAST_HPP_
+#define CONST_CAST_HPP_
+
+#include "openthread-core-config.h"
```

Comment:
Does this util depend on core config?

Reply:
It does not currently. But we have this in our style guide for all private/core files (which I think are very good rules to follow to ensure configs are always included).

>   - Preprocessor `#include` directives should be grouped, ordered, or sorted as follows:
>     - **If the unit is a core/private header file, `"openthread-core-config.h"` should be the first header file included.**
>     - If the unit is a core/private `.c` or `.cpp` file:
>       - If the unit has a corresponding header file, the unit's corresponding header file should be included before any other header file.
>       - If the unit has no corresponding header file, then it should directly include `"openthread-core-config.h"` before any other header file.

Reply:
These code kind of utilities are not only used by core code but also platform code and libraries. I don't think they are real core/private headers. Perhaps we should move them out of core. Anyway, I'm fine with the PR as is.